File: static/js/sync.js
/* Amazon <-> App sync UI (feature 2).
   Pull = Amazon -> app; Push = app -> Amazon. Both PROPOSE a side-by-side; nothing is
   overwritten without a per-field choice. Push buttons are locked until unlocked, and the
   backend refuses the actual write (halted + borrowed-cred refusal). Gated on the live
   capability matrix. Status pills show "Unknown -- never pulled" before a first pull. */

let SYNC_PUSH_UNLOCKED = false;
let SYNC_LAST = null;   // {sku, amazon, stored, status} stashed on a pull, for apply

function _syncToast(msg){
  var t=document.getElementById('toast'); if(!t){ return; }
  t.textContent=msg; t.classList.add('show');
  clearTimeout(window._syncTt); window._syncTt=setTimeout(function(){ t.classList.remove('show'); }, 3200);
}

function _syncPill(status){
  var m={
    confirmed:['#1f3d2e','#4caf7d','Authorised'],
    synced:['#1f3d2e','#4caf7d','Synced'],
    'unknown-never-pulled':['#33384a','#aab3c5','Unknown — never pulled'],
    'regenerated-not-on-amazon':['#3a3320','#e3b768','Regenerated — not on Amazon'],
    'changed-on-amazon':['#1e3350','#6ea8fe','Changed on Amazon — app behind'],
    deactivated:['#3d1f22','#e0696b','Deactivated (temporary)'],
    role_gap:['#3d1f22','#e0696b','No read role'],
    blocked_unconfirmed:['#3a3320','#e3b768','Blocked — reason unconfirmed'],
    borrowed:['#33384a','#aab3c5','Borrowed (catalogue-only)'],
    not_connected:['#33384a','#aab3c5','Not connected'],
    untested:['#33384a','#aab3c5','Not read-tested']
  };
  var p=m[status]||['#33384a','#aab3c5',status||'?'];
  return '<span style="background:'+p[0]+';color:'+p[1]+';padding:2px 9px;border-radius:10px;font-size:11px;font-weight:600;white-space:nowrap">'+esc(p[2])+'</span>';
}

function syncOnOpen(){ syncRenderMatrix(); }

async function syncRenderMatrix(){
  var el=document.getElementById('sync_matrix'); if(!el) return;
  el.innerHTML='<span class="genspin"></span> Loading capabilities…';
  try{
    var j=await (await fetch('/sync/capabilities')).json();
    if(!j.ok){ el.innerHTML='<div class="cc" style="color:#e0696b">could not load</div>'; return; }
    var h='<table class="ishtable" style="width:100%"><thead><tr><th>Account</th><th>Status</th><th>Pull</th><th>Push</th></tr></thead><tbody>';
    (j.accounts||[]).forEach(function(a){
      h+='<tr><td><b>'+esc(a.label)+'</b></td>'
        +'<td>'+_syncPill(a.status)+'</td>'
        +'<td title="'+esc(a.reason)+'">'+(a.pull_enabled?'<span style="color:#4caf7d">active</span>':'<span style="color:#e0696b">disabled</span>')+'</td>'
        +'<td>'+(a.push_enabled?'<span style="color:#e3b768">inferred</span>':'<span style="color:#8b93a5">—</span>')+'</td></tr>';
    });
    h+='</tbody></table>'
      +'<div class="cc" style="font-size:11px;opacity:.7;margin-top:6px">Pull reflects a live read-test. Push is inferred until a real push. Re-check / mark-cause apply to the <b>active workspace</b>.</div>'
      +'<div style="margin-top:10px;display:flex;gap:8px;flex-wrap:wrap">'
      +'<button class="btn" onclick="syncRecheck()">Re-check active account (read-test)</button>'
      +'<button class="btn" onclick="syncMarkCause()">Mark cause…</button></div>';
    el.innerHTML=h;
  }catch(e){ el.innerHTML='<div class="cc" style="color:#e0696b">error loading capabilities</div>'; }
}

function _syncSku(){ var e=document.getElementById('sync_sku'); return e?(e.value||'').trim():''; }

async function syncCheckStatus(){
  var sku=_syncSku(); var pill=document.getElementById('sync_pill');
  if(!sku){ pill.innerHTML=''; return; }
  pill.innerHTML='<span class="genspin"></span>';
  try{
    var j=await (await fetch('/sync/status',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({sku:sku})})).json();
    pill.innerHTML = j.ok ? (_syncPill(j.status)+' <span class="cc" style="font-size:11px;opacity:.7">'+esc(j.detail||'')+'</span>')
                          : ('<span style="color:#e0696b">'+esc(j.error||'error')+'</span>');
  }catch(e){ pill.innerHTML='<span style="color:#e0696b">error</span>'; }
}

async function syncPull(){
  var sku=_syncSku(); if(!sku){ _syncToast('Enter a SKU'); return; }
  _syncToast('Pulling from Amazon…');
  try{
    var j=await (await fetch('/sync/pull',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({sku:sku})})).json();
    if(!j.ok){ _syncToast(j.error||'pull failed'); return; }
    syncOpenModal('pull', sku, j.stored_copy||{}, j.amazon_copy||{}, j);
  }catch(e){ _syncToast('pull error'); }
}

async function syncPush(){
  if(!SYNC_PUSH_UNLOCKED){ _syncToast('Push is locked — tick "Unlock push" first.'); return; }
  var sku=_syncSku(); if(!sku){ _syncToast('Enter a SKU'); return; }
  try{
    var j=await (await fetch('/sync/push',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({sku:sku})})).json();
    if(!j.ok){ _syncToast(j.error||'push blocked'); return; }
    syncOpenModal('push', sku, j.stored_copy||{}, j.amazon_copy||{}, j);
  }catch(e){ _syncToast('push error'); }
}

var _SYNC_FIELDS=[['title','Title'],['item_highlights','Item Highlights'],['bullet_1','Bullet 1'],
  ['bullet_2','Bullet 2'],['bullet_3','Bullet 3'],['bullet_4','Bullet 4'],['bullet_5','Bullet 5'],
  ['description','Description'],['search_terms','Search terms']];

function syncOpenModal(mode, sku, stored, amazon, meta){
  SYNC_LAST={sku:sku, amazon:amazon, stored:stored, status:(meta&&meta.status)};
  var h='<div class="cc" style="margin-bottom:8px">'+esc(sku)+' — '
    +(mode==='pull'?'PULL: tick the Amazon fields to bring into the app.':'PUSH: review what would go to Amazon (write is halted).')+'</div>';
  if(meta&&meta.warn){ h+='<div style="background:#3a3320;color:#e3b768;padding:8px 10px;border-radius:8px;margin-bottom:8px">'+esc(meta.warn)+'</div>'; }
  if(meta&&meta.status){ h+='<div style="margin-bottom:8px">Status: '+_syncPill(meta.status)+'</div>'; }
  var srcLbl=(mode==='pull')?'Amazon (incoming)':'App (would push)';
  var tgtLbl=(mode==='pull')?'App (current)':'Amazon (current)';
  h+='<table class="ishtable" style="width:100%;table-layout:fixed"><thead><tr>';
  if(mode==='pull') h+='<th style="width:34px"></th>';
  h+='<th style="width:120px">Field</th><th>'+esc(srcLbl)+'</th><th>'+esc(tgtLbl)+'</th></tr></thead><tbody>';
  _SYNC_FIELDS.forEach(function(fp){
    var f=fp[0], lbl=fp[1];
    var src=(mode==='pull'?amazon:stored)[f]||'';
    var tgt=(mode==='pull'?stored:amazon)[f]||'';
    var diff=(src!==tgt);
    h+='<tr style="'+(diff?'background:rgba(110,168,254,.06)':'')+'">';
    if(mode==='pull') h+='<td>'+(diff?'<input type="checkbox" class="sync_apply" data-f="'+f+'">':'')+'</td>';
    h+='<td class="cc" style="font-size:11px">'+esc(lbl)+(diff?' <span style="color:#6ea8fe">●</span>':'')+'</td>'
      +'<td style="font-size:11px;word-break:break-word">'+esc(String(src).slice(0,300))+'</td>'
      +'<td style="font-size:11px;word-break:break-word">'+esc(String(tgt).slice(0,300))+'</td></tr>';
  });
  h+='</tbody></table>';
  if(mode==='pull'){
    h+='<div style="margin-top:12px"><button class="btn primary" onclick="syncApplyPull()">Apply checked Amazon fields to app</button></div>';
  }else{
    h+='<div style="margin-top:12px;background:#3d1f22;color:#e0696b;padding:8px 10px;border-radius:8px">Push write is intentionally HALTED pending the reviewed one-listing first-push test. Nothing was sent to Amazon.</div>';
  }
  document.getElementById('sync_modal_title').textContent=(mode==='pull'?'Pull review':'Push review')+' — '+sku;
  document.getElementById('sync_modal_body').innerHTML=h;
  document.getElementById('sync_modal').classList.add('show');
}
function syncCloseModal(){ document.getElementById('sync_modal').classList.remove('show'); }

async function syncApplyPull(){
  if(!SYNC_LAST){ return; }
  var fields={};
  document.querySelectorAll('#sync_modal_body .sync_apply:checked').forEach(function(cb){
    fields[cb.dataset.f]=(SYNC_LAST.amazon[cb.dataset.f]||'');
  });
  if(!Object.keys(fields).length){ _syncToast('Tick at least one field'); return; }
  var body={sku:SYNC_LAST.sku, fields:fields};
  try{
    var r=await fetch('/sync/pull/apply',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(body)});
    var j=await r.json();
    if(r.status===409 && j.blocked){
      if(!confirm((j.error||'This row was regenerated but not pushed.')+'\n\nApply anyway and DISCARD the regenerated copy?')){ return; }
      body.force=true;
      j=await (await fetch('/sync/pull/apply',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(body)})).json();
    }
    _syncToast(j.ok?('Applied to app: '+((j.applied||[]).join(', ')||'nothing')):(j.error||'apply failed'));
    if(j.ok){ syncCloseModal(); syncCheckStatus(); }
  }catch(e){ _syncToast('apply error'); }
}

async function syncRecheck(){
  _syncToast('Re-testing active account…');
  try{
    var j=await (await fetch('/sync/read_test',{method:'POST',headers:{'Content-Type':'application/json'},body:'{}'})).json();
    _syncToast(j.ok?('Pull: '+j.status+(j.pull_enabled?' (active)':' (disabled)')):(j.error||'read-test failed'));
    syncRenderMatrix();
  }catch(e){ _syncToast('read-test error'); }
}

async function syncMarkCause(){
  var c=prompt('Mark the pull-block cause for the ACTIVE workspace. Type one:\n'
    +'  deactivated          (temporary suspension -- clears on reinstatement)\n'
    +'  role_gap             (credentials genuinely lack the read role)\n'
    +'  blocked_unconfirmed  (reset -- reason unknown)');
  if(!c) return;
  var st=c.trim();
  if(['deactivated','role_gap','blocked_unconfirmed','untested'].indexOf(st)<0){ _syncToast('invalid status'); return; }
  var note=prompt('Optional note (e.g. "Section 3 suspension"):')||'';
  try{
    var j=await (await fetch('/sync/mark_status',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({status:st,note:note})})).json();
    _syncToast(j.ok?'marked':(j.error||'failed')); syncRenderMatrix();
  }catch(e){ _syncToast('mark error'); }
}

function syncTogglePushLock(cb){
  SYNC_PUSH_UNLOCKED=!!cb.checked;
  var b=document.getElementById('sync_push_btn');
  if(b){ b.disabled=!cb.checked; b.title=cb.checked?'Push (backend write halted until the first-push test)':'Push is locked'; }
}

File: listing/sync.py
"""listing/sync.py -- two-way Amazon <-> app sync core (feature 2). SAFE BY DESIGN.

Pull = Amazon -> app; Push = app -> Amazon. These are OPPOSITE directions and are kept
strictly separate. NOTHING here writes blindly: pull/push functions PROPOSE (return both
versions) and only the *apply/confirm* functions write, one listing at a time. A row that
was regenerated but not yet pushed is protected from being overwritten by a pull.

Capability is gated per account: PULL is only offered where a live read-test confirmed it;
PUSH is only offered where the account can publish (still INFERRED until a real push).

Snapshot model (answers "what does status show before a first pull?"):
  * The per-row status is computed against a CACHED last-known-Amazon snapshot
    (sync_snapshots.json, in the data dir), which is populated ONLY by a pull.
  * Before any pull has ever run for a SKU there is NO snapshot, so status is
    "unknown-never-pulled" -- NEVER a false "synced". Honest unknown, not a guess.
"""
import os
import json
import time

import domain.accounts as _acc

# Data caches live beside config.json (the data dir), never with the code.
def _data_dir():
    import amazon_listing_generator as G
    return str(G.CONFIG_PATH.parent)

_SNAP_FILE = "sync_snapshots.json"       # {account_id: {sku: {"copy":{...}, "pulled_at":ts}}}
_CAP_FILE  = "sync_capability.json"      # {account_id: {"pull_confirmed":bool, "tested_at":ts}}

COPY_FIELDS = ["title", "item_highlights", "bullet_1", "bullet_2", "bullet_3",
               "bullet_4", "bullet_5", "description", "search_terms"]


# --- small json cache helpers ----------------------------------------------
def _load(fn):
    try:
        return json.load(open(os.path.join(_data_dir(), fn), encoding="utf-8"))
    except Exception:
        return {}

def _save(fn, obj):
    try:
        json.dump(obj, open(os.path.join(_data_dir(), fn), "w", encoding="utf-8"),
                  ensure_ascii=False, indent=2)
        return True
    except Exception:
        return False


# --- capability (config-derived + live-confirmed pull) ---------------------
# Read-test outcomes. A FAILURE is not automatically a permanent 'no access': a 403 can
# mean the account is DEACTIVATED (temporary -- access returns on reinstatement) OR that
# the credentials genuinely lack the role (a real gap). Amazon's generic 403 does not say
# which, so an ambiguous failure is 'blocked_unconfirmed', never a guessed cause.
CONFIRMED = "confirmed"
DEACTIVATED = "deactivated"
ROLE_GAP = "role_gap"
BLOCKED_UNCONFIRMED = "blocked_unconfirmed"
UNTESTED = "untested"


def classify_read_error(err_text: str) -> str:
    """Classify a read-test failure from Amazon's message. Only auto-labels a cause when
    the message clearly says so; otherwise 'blocked_unconfirmed' (never a guess)."""
    m = (err_text or "").lower()
    if any(k in m for k in ("suspend", "deactivat", "inactive", "not active", "account status")):
        return DEACTIVATED
    if any(k in m for k in ("not authorized to perform", "grantless", "developer", "application id",
                            "missing role", "unauthorized for operation")):
        return ROLE_GAP
    return BLOCKED_UNCONFIRMED           # e.g. generic "Access to requested resource is denied"


def capability(account: dict) -> dict:
    """Per-account sync capability. PULL is enabled ONLY on a live-CONFIRMED read-test. A
    failure is labelled BY CAUSE (deactivated=temporary vs role_gap=real vs unconfirmed),
    and is always RE-TESTABLE -- one re-test on reinstatement flips it back to confirmed
    with no code change. PUSH is enabled from can_publish but stays INFERRED. An operator
    override can name a cause the generic 403 cannot."""
    aid = account.get("id", "")
    own = _acc.has_own_creds(account)
    borrowed = _acc.is_borrowed(account)
    can_read_scope = _acc.seller_scope_allowed(account)
    can_push_cfg = _acc.can_publish(account)
    cap = _load(_CAP_FILE).get(aid, {})
    status = cap.get("status")
    manual = cap.get("manual", "")
    if borrowed:
        status_lbl, reason = "borrowed", "borrowed credentials -> catalogue-only (no seller read/write)"
    elif not own:
        status_lbl, reason = "not_connected", "not connected (no own SP-API app)"
    elif status == CONFIRMED:
        status_lbl, reason = CONFIRMED, "own creds; pull CONFIRMED (live read-test)"
    elif status == DEACTIVATED:
        status_lbl, reason = DEACTIVATED, ("access blocked -- account DEACTIVATED (temporary); "
                                           "access returns on reinstatement, then re-test")
    elif status == ROLE_GAP:
        status_lbl, reason = ROLE_GAP, "credentials lack the Listings read role -- a real gap (fix creds)"
    elif status == BLOCKED_UNCONFIRMED:
        status_lbl, reason = BLOCKED_UNCONFIRMED, ("access blocked -- reason UNCONFIRMED (generic 403); "
                                                   "re-test, or mark the cause (deactivated vs role gap)")
    else:
        status_lbl, reason = UNTESTED, "own creds; pull not yet read-tested"
    if manual:
        reason += " | operator note: " + manual
    pull_confirmed = (status == CONFIRMED)
    return {
        "account_id": aid,
        "status": status_lbl,
        "pull_enabled": bool(can_read_scope and pull_confirmed),
        "pull_possible": bool(can_read_scope),
        "pull_confirmed": pull_confirmed,
        "re_testable": bool(own and not borrowed),      # a failure is never a permanent verdict
        "push_enabled": bool(can_push_cfg),              # INFERRED until a real push
        "push_confirmed": bool(cap.get("push_confirmed")),
        "reason": reason,
        "tested_at": cap.get("tested_at"),
    }


def set_pull_result(account_id: str, status: str, detail: str = ""):
    """Record a read-test outcome. Re-running the read-test overwrites this -- never frozen.
    A fresh CONFIRMED clears any stale operator note (access is back)."""
    cap = _load(_CAP_FILE)
    e = cap.setdefault(account_id, {})
    e["status"] = status
    e["detail"] = detail
    e["tested_at"] = int(time.time())
    if status == CONFIRMED:
        e.pop("manual", None)
    _save(_CAP_FILE, cap)


def mark_status(account_id: str, status: str, note: str = ""):
    """OPERATOR override to name a cause the 403 cannot (e.g. 'deactivated'). Sets status +
    a note; does NOT touch the Amazon account. A later successful re-test clears it."""
    cap = _load(_CAP_FILE)
    e = cap.setdefault(account_id, {})
    e["status"] = status
    e["manual"] = note
    e.setdefault("tested_at", int(time.time()))
    _save(_CAP_FILE, cap)


# --- per-row status --------------------------------------------------------
UNKNOWN   = "unknown-never-pulled"
SYNCED    = "synced"
APP_AHEAD = "regenerated-not-on-amazon"
AMZ_AHEAD = "changed-on-amazon"

def _snapshot(account_id, sku):
    return _load(_SNAP_FILE).get(account_id, {}).get(sku)

def compute_status(account_id: str, sku: str, stored_copy: dict, regenerated_at) -> dict:
    """Honest per-row status. No snapshot yet -> UNKNOWN (never a false 'synced')."""
    snap = _snapshot(account_id, sku)
    if not snap:
        return {"status": UNKNOWN, "label": "Unknown -- never pulled",
                "detail": "no Amazon snapshot yet; run Pull to establish the true state"}
    amazon = snap.get("copy", {})
    pulled_at = snap.get("pulled_at", 0)
    # app has regenerated since the last pull -> app is ahead of Amazon
    if regenerated_at and regenerated_at > pulled_at:
        return {"status": APP_AHEAD, "label": "Regenerated -- not yet on Amazon",
                "detail": "app copy is newer than the last pull; push (with review) to publish"}
    # stored copy differs from the last known Amazon copy, with no app-side change
    if {k: stored_copy.get(k, "") for k in COPY_FIELDS} != {k: amazon.get(k, "") for k in COPY_FIELDS}:
        return {"status": AMZ_AHEAD, "label": "Changed on Amazon -- app behind",
                "detail": f"differs from last pull ({time.strftime('%Y-%m-%d %H:%M', time.localtime(pulled_at))}); pull to review"}
    return {"status": SYNCED, "label": "Synced",
            "detail": f"matches Amazon as of last pull ({time.strftime('%Y-%m-%d %H:%M', time.localtime(pulled_at))})"}


# --- pull (Amazon -> app): PROPOSE only, plus snapshot update --------------
def _amazon_copy_from_item(pay: dict) -> dict:
    """Extract the copy fields we track from a getListingsItem payload."""
    out = {k: "" for k in COPY_FIELDS}
    summ = (pay or {}).get("summaries") or []
    if summ:
        out["title"] = summ[0].get("itemName", "") or ""
    attrs = (pay or {}).get("attributes") or {}
    def _first(key):
        v = attrs.get(key)
        if isinstance(v, list) and v and isinstance(v[0], dict):
            return v[0].get("value", "")
        return ""
    for i in range(1, 6):
        bp = attrs.get("bullet_point") or []
        if isinstance(bp, list) and len(bp) >= i and isinstance(bp[i-1], dict):
            out[f"bullet_{i}"] = bp[i-1].get("value", "")
    out["description"] = _first("product_description")
    return out


def pull_from_amazon(cfg: dict, account: dict, sku: str) -> dict:
    """LIVE READ. Fetch Amazon's current copy for one SKU, update the snapshot, and
    RETURN it for a side-by-side (does NOT write to the sheet). Caller applies chosen
    fields via the /apply endpoint. Returns {"ok","amazon_copy","error"}."""
    cap = capability(account)
    if not cap["pull_enabled"]:
        return {"ok": False, "error": "pull not enabled for this account: " + cap["reason"]}
    try:
        from sp_api.api import ListingsItemsV20210801 as LI
        from sp_api.base import Marketplaces
    except Exception as e:
        return {"ok": False, "error": f"sp_api unavailable: {e}"}
    mkt = (account.get("default_marketplace") or "UK").upper()
    mkt = "UK" if mkt == "GB" else mkt
    mid = _acc.marketplace_id(mkt)
    creds = _acc.account_creds(account)
    try:
        li = LI(credentials=creds, marketplace=getattr(Marketplaces, mkt, Marketplaces.UK), timeout=60)
        resp = li.get_listings_item(account.get("seller_id", ""), sku,
                                    marketplaceIds=[mid] if mid else None,
                                    includedData="summaries,attributes")
        pay = resp.payload if hasattr(resp, "payload") else (resp or {})
    except Exception as e:
        return {"ok": False, "error": f"{type(e).__name__}: {str(e)[:160]}"}
    amazon = _amazon_copy_from_item(pay)
    snaps = _load(_SNAP_FILE)
    snaps.setdefault(account.get("id", ""), {})[sku] = {"copy": amazon, "pulled_at": int(time.time())}
    _save(_SNAP_FILE, snaps)
    return {"ok": True, "amazon_copy": amazon}


# --- push (app -> Amazon): PROPOSE + gated confirm -------------------------
def push_to_amazon(cfg: dict, account: dict, sku: str, fields: dict) -> dict:
    """LIVE WRITE (patchListingsItem). GATED: the caller must have shown the side-by-side
    and the operator confirmed. This is the first-push test surface -- run ONE listing,
    review, report, before any second push. Returns {"ok","error"}.

    NOTE: enforcement of the business rules (LISTING requirements, GTIN exemption, brand
    kept, never merchant_suggested_asin / LISTING_OFFER_ONLY) is the generator's submit
    path -- this function is intentionally a stub that REFUSES until wired to that path,
    so 'push is wired' can never silently become 'push wrote to a live listing'."""
    cap = capability(account)
    if not cap["push_enabled"]:
        return {"ok": False, "error": "push not enabled for this account: " + cap["reason"]}
    # SAFETY STOP: no live write is performed here yet. The first live push must be an
    # explicit, reviewed, one-listing test wired through the generator's compliant submit
    # path -- deliberately not auto-executed.
    return {"ok": False, "error": "push path wired but LIVE WRITE is intentionally halted "
                                  "pending the gated first-push test (one listing, reviewed)."}
